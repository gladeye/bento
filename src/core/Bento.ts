import { resolve, basename } from "path";
import {
    Loader,
    Condition,
    Configuration,
    Resolve,
    Entry,
    Plugin
} from "webpack";
import { sync as resolveModuleSync } from "resolve";
import { isString, isArray } from "lodash";
import { instantiate, selector } from "../utils/lang";

export interface Config {
    cwd?: string;
    homeDir: string;
    outputDir: string;
    publicPath?: string;
}

export interface RuleMap {
    [ext: string]: Loader | Loader[];
}

export interface RuleDescriptor {
    ext: string[];
    include: Condition | Condition[];
    loaders: Loader[] | ((env: string | void) => Loader | Loader[]);
}

export interface PluginMap {
    [name: string]: any[];
}

export interface PluginDescriptor {
    plugin: string | Constructor<Plugin>;
    args: any[] | ((env?: string) => any[] | void);
}

export interface PluginCollection {
    [env: string]: PluginDescriptor[];
}

export interface Features {
    sourceMap: boolean;
}

export enum Env {
    Production = "production"
}

export enum Command {
    Bundle = "webpack",
    Serve = "webpack-dev-server"
}

interface Constructor<M> {
    new (...args: any[]): M;
}

/**
 * Main class of the library
 *
 * @export
 * @class Bento
 */
export default class Bento {
    /**
     * @private
     * @type {Config}
     * @memberof Bento
     */
    protected config: Config;

    /**
     * @private
     * @type {RuleDescriptor[]}
     * @memberof Bento
     */
    protected rules: RuleDescriptor[] = [];

    /**
     * @private
     * @type {PluginCollection}
     * @memberof Bento
     */
    protected plugins: PluginCollection = {};

    /**
     * @private
     * @type {Entry}
     * @memberof Bento
     */
    protected entry: Entry = {};

    /**
     * @private
     * @type {Features}
     * @memberof Bento
     */
    protected features: Features = {
        sourceMap: true
    };

    /**
     * @private
     * @memberof Bento
     */
    protected resolve: (path: string) => string;

    /**
     * Creates an instance of Bento.
     * @param {Config} config
     * @memberof Bento
     */
    constructor(config: Config) {
        this.config = config;
        this.resolve = resolve.bind(this, this.cwd);

        this.load();
    }

    /**
     * Factory method to create an instance of Bento
     *
     * @static
     * @param {Config} config
     * @param {string} [cwd]
     * @returns {T}
     * @memberof T
     */
    static create<T extends Bento>(
        this: Constructor<T>,
        config: Config,
        cwd?: string
    ): T {
        return new this(config, cwd);
    }

    /**
     * Define `entry` value for webpack config
     *
     * @param {string} name
     * @param {...string[]} files
     * @returns {this}
     * @memberof Bento
     */
    bundle(name: string, files: string | string[]): this {
        if (isString(files)) files = [files as string];
        this.entry[name] = (files as string[]).map((file) =>
            file.replace("~/", `${this.config.homeDir}/`)
        );
        return this;
    }

    /**
     * Allow all configuration to be modified before webpack configuration
     * is built
     *
     * @param {(manifest: {}) => void} [fn]
     * @returns {this}
     * @memberof Bento
     */
    tinker(fn?: (manifest: {}) => void): this {
        const manifest = {
            rules: this.rules,
            plugins: this.plugins,
            entry: this.entry
        };

        fn(manifest);

        return this;
    }

    /**
     * Create a new RuleDescriptor by a given data and add it to rule list
     *
     * @param {string} ext  File extension
     * @param {Loader | Loader[]} loaders Loaders to use
     * @param {Condition | Condition[]} include Specify path inclusion
     * @returns {this}
     * @memberof Bento
     */
    addRule(
        ext: string | string[],
        loaders: Loader | Loader[] | ((env?: string) => Loader | Loader[]),
        include?: Condition | Condition[]
    ): this {
        if (isString(ext)) ext = [ext as string];
        if (typeof loaders !== "function") {
            if (!Array.isArray(loaders)) loaders = [loaders as Loader];
            loaders = loaders.map((loader) => {
                if (typeof loader === "string") return { loader };
                return loader;
            });
        }

        this.rules.push({
            ext: ext as string[],
            include: include || [this.homeDir],
            loaders
        });

        return this;
    }

    /**
     * Convenient method to add multiple rules
     *
     * @param {RuleMap} map Map of file extensions and their creators
     * @returns {this}
     * @memberof Bento
     */
    addRules(map: RuleMap = {}): this {
        for (let p in map) {
            this.addRule(p.replace(/\s+/g, "").split(","), map[p]);
        }

        return this;
    }

    /**
     * Add to plugin list a new PluginDescriptor by given data
     *
     * @param {string | Constructor<Plugin>} plugin Plugin constructor
     * @param {any[]} args Plugin constructor arguments
     * @param {string} env Env this plugin should be loaded to
     * @returns {this}
     * @memberof Bento
     */
    addPlugin(
        plugin: string | Constructor<Plugin>,
        args: any[] | ((env?: string) => any[] | void) = [],
        env?: string
    ): this {
        let key = env || "all";

        if (!this.plugins[key]) this.plugins[key] = [];

        this.plugins[key].push({
            plugin,
            args
        });

        return this;
    }

    /**
     * Convenient method to add multiple plugins
     *
     * @param {PluginMap} map
     * @param {string} env Env this plugin should be loaded to
     * @returns {this}
     * @memberof Bento
     */
    addPlugins(map: PluginMap = {}, env?: string): this {
        for (let p in map) {
            this.addPlugin(p, map[p], env);
        }

        return this;
    }

    /**
     * Export data to a standard consumable webpack config file
     *
     * @param {string | void} [env]
     * @returns {Promise<Configuration>}
     * @memberof Bento
     */
    export(env?: string | void): Promise<Configuration> {
        const select = selector(env);

        const plugins = [].concat(
            this.plugins.all || [],
            env ? this.plugins[env] || [] : []
        );

        type WebpackMode = "development" | "production" | "none";

        const config = {
            name: "bento",
            entry: this.entry,
            mode: (env === Env.Production
                ? "production"
                : "development") as WebpackMode,
            output: {
                path: this.outputDir,
                pathinfo: select({
                    default: true,
                    production: false
                }),
                filename: select({
                    default: "[name].js",
                    production: "[name].[chunkhash:8].js"
                }),
                publicPath: this.publicPath
            },
            devtool: select({
                default: this.features.sourceMap
                    ? "cheap-module-eval-source-map"
                    : null,
                production: this.features.sourceMap ? "source-map" : false
            }),
            resolve: {
                alias: {
                    "~": this.homeDir
                }
            },
            module: {
                rules: this.rules.map((desc) => {
                    return {
                        test: new RegExp(`\\.(${desc.ext.join("|")})$`),
                        include: desc.include,
                        use:
                            typeof desc.loaders === "function"
                                ? desc.loaders(env)
                                : desc.loaders.map((loader: any) => {
                                      return typeof loader === "function"
                                          ? loader(env)
                                          : loader;
                                  })
                    };
                })
            },
            plugins: plugins
                .map((desc) => {
                    let Plugin;
                    if (typeof desc.plugin === "function") Plugin = desc.plugin;
                    else {
                        const module = resolveModuleSync(desc.plugin, {
                            basedir: this.cwd
                        });

                        Plugin = require(module);
                    }

                    const args =
                        typeof desc.args === "function"
                            ? desc.args(env)
                            : desc.args;

                    return isArray(args) ? instantiate(Plugin, args) : null;
                })
                .filter(Boolean)
        };

        return Promise.resolve(this.configure(config, env));
    }

    /**
     * Enable / disable available features
     *
     * @param {string} flag
     * @param {any} value
     * @returns {this}
     * @memberof Bento
     */
    set(flag: string, value: any): this {
        this.features[flag] = value;
        return this;
    }

    /**
     * Find loader with a given `name`, `cb` will be triggered if found.
     *
     * @param {string} name
     * @param {({ loader, ext, include }) => void} cb
     * @memberof Bento
     */
    findLoader(name: string, cb: ({ loader, ext, include }) => void) {
        this.rules.forEach((rule) => {
            const trigger = (loader) => {
                if (loader.loader && loader.loader === name)
                    cb({ loader, ext: rule.ext, include: rule.include });
            };

            if (typeof rule.loaders === "function") return;
            rule.loaders.forEach(trigger);
        });
    }

    /**
     * Resolved homeDir
     *
     * @readonly
     * @protected
     * @type {string}
     * @memberof Bento
     */
    protected get homeDir(): string {
        return this.resolve(this.config.homeDir);
    }

    /**
     * Resolved outputDir
     *
     * @readonly
     * @protected
     * @type {string}
     * @memberof Bento
     */
    protected get outputDir(): string {
        return this.resolve(this.config.outputDir);
    }

    /**
     * Default publicPath
     *
     * @readonly
     * @protected
     * @type {string}
     * @memberof Bento
     */
    protected get publicPath(): string {
        return this.config.publicPath || "/";
    }

    /**
     * @readonly
     * @protected
     * @type {string}
     * @memberof Bento
     */
    protected get cwd(): string {
        return this.config.cwd || process.cwd();
    }

    /**
     * Placeholder method for subclass to add in rules & plugins
     *
     * @protected
     * @memberof Bento
     */
    protected load(): void {}

    /**
     * Placeholder method for subclass to overwrite `config`
     *
     * @protected
     * @param {Configuration} config
     * @param {string | void} env
     * @returns {Configuration}
     * @memberof Bento
     */
    protected configure(
        config: Configuration,
        env: string | void
    ): Configuration {
        return config;
    }

    /**
     * @readonly
     * @protected
     * @type {string}
     * @memberof StandardBento
     */
    protected get command(): string {
        return basename(process.argv[1]);
    }
}
