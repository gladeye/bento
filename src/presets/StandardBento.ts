import {
    DefinePlugin,
    Loader,
    Configuration,
    NamedModulesPlugin,
    NamedChunksPlugin
} from "webpack";
// import * as ExtractTextWebpackPlugin from "extract-text-webpack-plugin";
import * as MiniCssExtractPlugin from "mini-css-extract-plugin";
import * as UglifyJsPlugin from "uglifyjs-webpack-plugin";
import * as OptimizeCSSAssetsPlugin from "optimize-css-assets-webpack-plugin";
import { basename, extname } from "path";
import { isObject, isString } from "lodash";
import Bento, {
    Config as BaseConfig,
    Features as BaseFeatures,
    Command,
    Env
} from "../core/Bento";
import { selector } from "../utils/lang";

export interface Config extends BaseConfig {
    html?: string | {};
    proxy?: string | {};
}

export interface Features extends BaseFeatures {
    extractCss: boolean;
    emitFiles: boolean | RegExp;
}

export default class StandardBento extends Bento {
    /**
     * @protected
     * @type {Config}
     * @memberof StandardBento
     */
    protected config: Config;

    /**
     * @protected
     * @type {Features}
     * @memberof StandardBento
     */
    protected features: Features;

    /**
     * @see Bento#load
     * @protected
     * @memberof StandardBento
     */
    protected load() {
        // set default addtional features flag
        this.set("extractCss", true).set("emitFiles", false);

        // SCRIPT
        this.addRule(["js", "jsx"], {
            loader: "babel-loader",
            options: {
                presets: [
                    [
                        "@babel/preset-env",
                        {
                            useBuiltIns: "usage"
                        }
                    ]
                ]
            }
        })
            .addPlugin("clean-webpack-plugin", [
                this.config.outputDir,
                {
                    root: this.cwd,
                    verbose: false
                }
            ])
            .addPlugin("webpack-manifest-plugin", [
                {
                    filter(descriptor) {
                        const pathExt = extname(descriptor.path),
                            nameExt = extname(descriptor.name);

                        return pathExt === nameExt;
                    }
                }
            ])
            .addPlugin(
                "write-file-webpack-plugin",
                (env?: string): void | any[] => {
                    let args = null;
                    if (this.features.emitFiles === true) args = [];
                    if (this.features.emitFiles instanceof RegExp)
                        args = [
                            {
                                test: this.features.emitFiles
                            }
                        ];

                    return args;
                }
            )
            .addPlugin(NamedChunksPlugin, [
                (chunk) => {
                    if (chunk.name) {
                        return chunk.name;
                    }

                    return Array.from(chunk.modulesIterable)
                        .map((m: any) => {
                            return basename(m.request, extname(m.request));
                        })
                        .join("_");
                }
            ])
            .addPlugin(NamedModulesPlugin, [])
            .addPlugin(
                DefinePlugin,
                (env?: string): any[] => {
                    return [
                        {
                            "process.env": {
                                NODE_ENV: JSON.stringify(env)
                            }
                        }
                    ];
                }
            );
        // .addPlugin(
        //     optimize.CommonsChunkPlugin,
        //     [
        //         {
        //             name: "vendor",
        //             minChunks(module) {
        //                 return /node_modules/.test(module.resource);
        //             }
        //         }
        //     ],
        //     Env.Production
        // )
        // .addPlugin(
        //     optimize.CommonsChunkPlugin,
        //     [
        //         {
        //             name: "runtime"
        //         }
        //     ],
        //     Env.Production
        // )
        // .addPlugin(
        //     "uglifyjs-webpack-plugin",
        //     (env?: string): any[] => {
        //         return [
        //             {
        //                 sourceMap: this.features.sourceMap
        //             }
        //         ];
        //     },
        //     Env.Production
        // );

        // STYLE
        this.addRule(
            "scss",
            (env?: string): Loader[] => {
                const loaders = [
                    {
                        loader: "css-loader",
                        options: {
                            sourceMap: this.features.sourceMap,
                            minimize:
                                env === Env.Production
                                    ? { preset: "default" }
                                    : false
                        }
                    },
                    {
                        loader: "postcss-loader",
                        options: {
                            ident: "postcss",
                            plugins() {
                                const autoprefixer = require("autoprefixer");
                                return [autoprefixer()];
                            },
                            sourceMap: this.features.sourceMap
                        }
                    },
                    {
                        loader: "resolve-url-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    },
                    {
                        loader: "sass-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    }
                ];

                if (env === Env.Production && this.features.extractCss) {
                    loaders.unshift({
                        loader: MiniCssExtractPlugin.loader,
                        options: {
                            publicPath: "./",
                            sourceMap: this.features.sourceMap
                        }
                    });
                } else {
                    loaders.unshift({
                        loader: "style-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    });
                }

                return loaders;
            }
        )
            .addRule(
                "css",
                [
                    {
                        loader: "style-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    },
                    {
                        loader: "css-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    }
                ],
                /node_modules/
            )
            .addPlugin(
                MiniCssExtractPlugin,
                (env?: string): any[] => {
                    const select = selector(env);

                    return [
                        {
                            filename: select({
                                default: "[name].css",
                                production: "[name].[contenthash:8].css"
                            })
                        }
                    ];
                }
            );

        // FONTS & MEDIA
        this.addRule(
            [
                "ttf",
                "eot",
                "woff",
                "woff2",
                "jpeg",
                "jpg",
                "png",
                "gif",
                "svg",
                "mp4",
                "webm"
            ],
            (env?: string): Loader => {
                const select = selector(env);

                return {
                    loader: "file-loader",
                    options: {
                        name: select({
                            default: "[name].[ext]",
                            production: "[name].[hash:8].[ext]"
                        })
                    }
                };
            },
            [this.homeDir, /node_modules/]
        );

        // HTML
        if (this.html) {
            this.addPlugin("html-webpack-plugin", [this.html]);
        }
    }

    /**
     * @see Bento#configure
     * @protected
     * @memberof StandardBento
     */
    protected configure(
        config: Configuration,
        env: string | void
    ): Configuration {
        if (env === Env.Production) {
            if (!config.optimization) config.optimization = {};

            config.optimization.splitChunks = {
                cacheGroups: {
                    commons: {
                        test: /[\\/]node_modules[\\/]/,
                        name: "vendor",
                        chunks: "all"
                    }
                }
            };

            config.optimization.runtimeChunk = {
                name: "runtime"
            };

            config.optimization.minimizer = [
                new UglifyJsPlugin({
                    cache: true,
                    parallel: true,
                    sourceMap: this.features.sourceMap
                }),
                new OptimizeCSSAssetsPlugin({})
            ];
        }

        if (this.command !== Command.Serve) return config;

        config.devServer = {
            disableHostCheck: true,
            historyApiFallback: true,
            overlay: true
        };

        if (this.proxy) config.devServer.proxy = this.proxy;

        return config;
    }

    /**
     * @readonly
     * @protected
     * @type {({} | void)}
     * @memberof StandardBento
     */
    protected get html(): {} | void {
        let output = null;
        if (isString(this.config.html)) {
            output = {
                filename: this.config.html
            };
        } else if (isObject(this.config.html)) {
            output = this.config.html;
        }

        return output;
    }

    /**
     * @readonly
     * @protected
     * @type {({} | void)}
     * @memberof StandardBento
     */
    protected get proxy(): {} | void {
        let output = null;
        if (isString(this.config.proxy)) {
            output = {
                "**": this.config.proxy
            };
        } else if (isObject(this.config.proxy)) {
            output = this.config.proxy;
        }

        return output;
    }
}
