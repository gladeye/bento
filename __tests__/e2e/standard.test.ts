import { resolve as resolvePath } from "path";
import * as execa from "execa";
import { Configuration } from "webpack";
import * as express from "express";
import * as fetch from "node-fetch";
import StandardBento from "~/presets/StandardBento";
import WebpackController from "~/core/WebpackController";
import { Env } from "~/core/Bento";

describe("serve", () => {
    const serve = function(cb) {
        return new Promise((resolve, reject) => {
            const cliPath = resolvePath(
                __dirname,
                "../../node_modules/.bin/webpack-dev-server"
            );
            const fixturePath = resolvePath(__dirname, "./serve");
            const nodePath = execa.shellSync("which node").stdout;

            const proc = execa(
                nodePath,
                [cliPath, "--hot", "--inline", "--port", "8000"],
                {
                    cwd: fixturePath
                }
            );

            proc.stdout.on("data", cb.bind(null, proc));
            proc.stderr.on("data", function(data) {
                if (data.indexOf("DeprecationWarning") < 0) reject(data);
                else console.log(data.toString());
            });
            proc.on("exit", resolve);
            proc.on("error", reject);
        });
    };

    const proxy = function() {
        const server = express();
        server.get("/", (req, res) => {
            res.send("proxy says hi");
        });

        return server.listen(9000);
    };

    it("should work as expected", async () => {
        const server = proxy();

        await serve(async (proc, data) => {
            const done = function() {
                server.close();
                proc.kill("SIGINT");
            };

            const output = data.toString();
            if (!/Compiled successfully/.test(output)) return;

            const mainBody = await (await fetch(
                "http://localhost:8000"
            )).text();

            const proxyBody = await (await fetch(
                "http://localhost:8000/api"
            )).text();

            expect(mainBody).toContain("<title>Webpack App</title>");
            expect(mainBody).toContain(
                `<script type="text/javascript" src="/main.js"></script>`
            );

            expect(proxyBody).toContain("proxy says hi");

            done();
        });
    }, 120000);
});

describe("build", () => {
    const build = function(
        entry,
        env?: string,
        overwrite?: (config: Configuration) => any
    ) {
        const bento = StandardBento.create({
            homeDir: "./__tests__",
            outputDir: "/"
        }).bundle("main", `~/fixtures/${entry}.js`);

        const files = {};

        return bento.export(env).then((config) => {
            if (overwrite) overwrite(config);

            config.performance = { hints: false };

            config.devtool = false;
            config.cache = false;
            const controller = new WebpackController(config);
            const promise = controller.compile();

            controller.compiler.outputFileSystem.writeFile = function(
                name,
                content,
                callback
            ) {
                files[name] = content.toString("utf-8");
                callback();
            };

            return promise.then((stats) => {
                return { files, stats };
            });
        });
    };

    it("exports a correct `config` for webpack", () => {
        return build("main2").then(({ files, stats }) => {
            expect(Object.keys(files)).toEqual([
                "/RobotoMono-Regular.ttf",
                "/giphy.gif",
                "/nice.jpg",
                "/cat.gif",
                "/main.js",
                "/manifest.json"
            ]);
            const bundle = files["/main.js"];

            expect(bundle).toContain("function __webpack_require__(");
            expect(bundle).toContain("__webpack_require__(/*! ~/fixtures/a */");
            expect(bundle).toContain("__webpack_require__(/*! ./style.scss */");
            expect(bundle).toContain("./__tests__/fixtures/main2.js");
            expect(bundle).toContain("./__tests__/fixtures/a.js");
            expect(bundle).toContain("./__tests__/fixtures/b.js");
            expect(bundle).toContain(
                "./__tests__/fixtures/node_modules/m1/a.js"
            );
            expect(bundle).toContain(
                "./__tests__/fixtures/node_modules/m1/s.css"
            );
            expect(bundle).toContain("This is a");
            expect(bundle).toContain("This is b");
            expect(bundle).toContain("This is m1/a");
            expect(bundle).toContain("var x = 1;");
            expect(bundle).toContain("color: red;");
            expect(bundle).toContain("font-size: 12px;");
            expect(bundle).toContain("console.log(undefined);");

            expect(bundle).not.toContain("4: function(");
            // expect(bundle).not.toContain("window");
            expect(bundle).not.toContain("jsonp");

            expect(JSON.parse(files["/manifest.json"])).toMatchSnapshot();
        });
    });

    it("exports a correct `config` for webpack in production", () => {
        return build("main2", Env.Production).then(({ files, stats }) => {
            const keys = Object.keys(files);
            expect(keys).toEqual([
                "/nice.9c3c4150.jpg",
                "/cat.5082946a.gif",
                "/RobotoMono-Regular.a48ac416.ttf",
                "/giphy.a47e713b.gif",
                "/main.04d8cdd3.css",
                "/main.fcc39b91.js",
                "/runtime.ae3ec808.js",
                "/vendor.eb592cd8.css",
                "/vendor.9e956fd5.js",
                "/manifest.json"
            ]);

            expect(files["/main.fcc39b91.js"]).toContain(`return\"This is a\"`);
            expect(files["/main.04d8cdd3.css"]).toContain("color:red");
            expect(files["/main.04d8cdd3.css"]).toContain("@-webkit-keyframes");
            expect(files["/vendor.eb592cd8.css"]).toContain("font-size:12px");
            expect(files["/vendor.9e956fd5.js"]).toContain("lodash");
            expect(JSON.parse(files["/manifest.json"])).toMatchSnapshot();
        });
    });

    // ref: https://medium.com/webpack/predictable-long-term-caching-with-webpack-d3eee1d3fa31
    it("keeps `vendor` chunk hash as static as possible", async () => {
        const find = (name, files) => {
            return Object.keys(files)
                .filter((item) => {
                    return item.indexOf(`/${name}`) === 0;
                })
                .shift();
        };

        // default
        const a = await build("main3-a", Env.Production);
        // change to entry
        const b = await build("main3-b", Env.Production);
        // load additional module async
        const c = await build("main3-c", Env.Production);
        // define `externals`
        const d = await build("main3-d", Env.Production, (config) => {
            config.externals = {
                jquery: "jQuery"
            };
            return config;
        });

        expect(find("vendor", b.files)).toBe(find("vendor", a.files));
        expect(find("vendor", c.files)).toBe(find("vendor", a.files));
        expect(find("vendor", d.files)).toBe(find("vendor", a.files));
    });
});
