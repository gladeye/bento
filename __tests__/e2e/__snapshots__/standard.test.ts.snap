// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build exports a correct \`config\` for webpack 1`] = `
Object {
  "RobotoMono-Regular.ttf": "/RobotoMono-Regular.ttf",
  "cat.gif": "/cat.gif",
  "giphy.gif": "/giphy.gif",
  "main.js": "/main.js",
  "nice.jpg": "/nice.jpg",
}
`;

exports[`build exports a correct \`config\` for webpack in production 1`] = `
Object {
  "cat.gif": "/cat.5082946a.gif",
  "main.css": "/main.04d8cdd3.css",
  "main.js": "/main.fcc39b91.js",
  "nice.jpg": "/nice.9c3c4150.jpg",
  "runtime.js": "/runtime.ae3ec808.js",
  "vendor.css": "/vendor.eb592cd8.css",
  "vendor.js": "/vendor.9e956fd5.js",
}
`;
